pdftops.1
.\" Copyright 1996 Derek B. Noonburg
.TH pdftops 1 "22 Feb 1998"
.SH NAME
pdftops \- Portable Document Format (PDF) to PostScript converter
(version 0.7a)
.SH SYNOPSIS
.B pdftops
[options]
.RI [ PDF-file
.RI [ PS-file ]]
.SH DESCRIPTION
.B Pdftops
converts Portable Document Format (PDF) files to PostScript so they
can be printed.
.PP
Pdftops reads the PDF file,
.IR PDF-file ,
and writes a PostScript file,
.IR PS-file .
If
.I PS-file
is not specified, pdftops converts
.I file.pdf
to
.IR file.ps .
If 
.I PS-file
is \'-', the PostScript is sent to stdout.
.PP
The generated PostScript is Level 2.  This means it will not work on
old printers which only support Level 1 PostScript.
.SH OPTIONS
.TP
.BI \-f number
Specifies the first page to print.
.TP
.BI \-l number
Specifies the last page to print.
.TP
.B \-level1
Generate Level 1 PostScript.  The resulting PostScript files will be
significantly larger (if they contain images), but will print on Level
1 printers.  This also converts all images to black and white.
.TP
.B \-noemb
By default, any Type 1 fonts which are embedded in the PDF file are
copied into the PostScript file.  This option causes pdftops to
substitute base fonts instead.  Embedded fonts make PostScript files
larger, but may be necessary for readable output.
.TP
.B \-h
Print usage information.
.RB ( \-help
is equivalent.)
.SH BUGS
This is a beta release of pdftops, and there are still a bunch of PDF
features which I have yet to implement.
.SH AUTHOR
The pdftops software and documentation are copyright 1996-1998 Derek
B. Noonburg (derekn@aimnet.com).
.SH "SEE ALSO"
.BR xpdf (1),
.BR pdftotext (1)
.br
.B http://www.aimnet.com/~derekn/xpdf/


CHANGES
0.2 (95-dec-12)
---------------
First public release.


0.3 (96-jan-13)
---------------
LZW patent workaround.
Implemented inline images.
Fixed (mostly) disjoint polygon fills.
Added remote server stuff.
Added page number on command line.
Fixed problem with font encodings which caused character misalignment.
Fixed inverted CCITT decoding and inverted image mask drawing.
Now compiles under gcc 2.7.x (ignore those stupid 'unused parameter'
  warnings).
Many minor bug fixes and optimizations.


0.4 (96-apr-24)
---------------
Implemented DCT filter.
Implemented PostScript output; wrote pdftops program.
Implemented links.
Implemented font rotation -- I was wrong: X11R6 servers *do* support
  font rotation (by specifying a matrix in place of a size).
Added bindings for Home/End, Page Up/Down, arrow keys.
Added initialZoom resource and -z option.
Added geometry resource and -g option.
Fixed image size off-by-one bug.
Fixed bug where page content is reference to an array of streams.
Cleaned up uninitialized variables which were causing random problems
  on various platforms.
Manually skip whitespace before calling atoi() for startxref.
Replaced calls to XrmCombineFileDatabase() with calls to
  XrmGetFileDatabase() and XrmMergeDatabases() so it will work under
  older versions of X.
Fixed problem with reading multiple xref tables in updated PDF files.
Check for encryption and print appropriate error message.
Rudimentary dithering of images.
Fixed bug in CCITTFax filter (pass mode followed by horizontal mode).
Optimized drawImage() and drawImageMask().
Changed several things to ease porting:
  - changed '__inline' to 'inline' (to adhere to the ANSI standard)
  - surrounded interface/implementation pragmas with #ifdef _GNUC__
  - got rid of empty array initializer lists
  - moved Operator type definition from Gfx.cc to Gfx.h
  - renamed String, uint, etc.
  - ability to uncompress to file (NO_POPEN and USE_GZIP flags)
  - added definitions of XK_Page_Up/Down and XPointer for old versions
    of X
For VMS port:
  - use correct Xdefaults name for VMS, get rid of ltkGetHomeDir()
  - added '#include <stddef.h>' before all X includes
  - renamed files with multiple periods in their names
Fixed window resizing infinite oscillation bug.
Fixed problem with string-type (as opposed to stream-type) indexed
  color space lookup tables (which are used in inline images).
If an X font is not found, try smaller and then larger sizes (this is
  useful for old or broken X servers which can't scale bitmap fonts).
Added -rgb (color cube size) option.
Various minor bug fixes.


0.5 (96-may-23)
---------------
Fixed bug in LTKWindow which broke the remote server mode.
Fixed PostScript output:
  - doesn't seg fault if file is unwritable.
  - incorrect DSC comment - need colon in '%%Page:'.
  - use 'imagemask' command for masks.
  - output filters in the correct order.
  - Stream::isBinary() checks the next-to-bottom, not top, stream.
  - if page width > height, rotate it 90 degrees.
  - if page is larger than paper size, scale it down.
Set default MediaBox to 8.5" x 11" to deal with non-compliant PDF
  files which don't specify a MediaBox.
Added DEBUG_MEM stuff in gmem.c and gmempp.cc.
Fixed memory leaks:
  - LTKWindow didn't delete the LTKBox.
  - LinkAction needs a virtual destructor.
Use $(RANLIB) variable in goo/Makefile and ltk/Makefile.
Allocate image data after calling XCreateImage, using
  image->bytes_per_line -- works in 24-bit mode now.
DCTStream rounds width of rowBuf lines up to the next multiple of
  mcuWidth, so last MCU doesn't run off end of buffer.
Increase size of block (from 255 to 1024 bytes) read at end of file to
  search for 'startxref'.
Skip past garbage at start of file, look for '%PDF'.
Moved more compiler options out of Makefiles into Makefile.config.
Top-level Makefile uses '$(MAKE)' instead of 'make' for making
  subdirectories.
Space/PageDown/Next and Backspace/PageUp/Previous now moves to
  next/previous page if already scrolled to bottom/top of current
  page.


0.5a (96-jul-09)
----------------
[not a public release]
For PDF 1.2 (a.k.a. Amber, a.k.a. Acrobat 3) support:
  - look for trailer after first xref instead of at end of file.
Deal with font subsets by converting character names of the form
  'Cnnnn' to the appropriate character from the standard encoding.
Extract encoding from embedded Type 1 fonts.
Kludge to fill one-pixel thick polygons.
Changed X font encoding to use endash for hyphen (gets rid of too-long
  hyphens).
Handle Resources key in Pages dictionaries (needed for pstoedit
  output).
Fix comment handling in Parser (needed for pstoedit output).
Move Bezier curve conversion from GfxState to XOutputDev; look at
  flatness parameter in GfxState.
Change all of the path functions in XOutputDev (stroke, fill, clip) to
  use the same path transformation/conversion function.
Rewrote PostScript output driver as a subclass of OutputDev; removed
  duplicated code (ps_ functions) from Gfx.
Fixed bug in xref code with small (< 1024 bytes) PDF files.
Implemented BX/EX operators.
Added PDFDoc class.


0.6 (96-nov-12)
---------------
Add support for PostScript output to stdout (-) and to a command
  (|lpr); added -ps option and psFile resource.
Decryption is implemented but not included in the distribution due to
  legal restrictions: the decryption algorithm is a trade secret of
  RSA, Inc., and the U.S.A. still has bogus export controls on
  cryptography software.
Added .xpdfrc config file:
  - Added fontmap parameter: user can map PDF font names to X fonts.
  - Added fontpath parameter: search for Type 1 font if encoding is
    not in PDF file.
Incremental display: display is updated after every 200 commands.
Added forward and backward by-10-page buttons.
Links:
  - Implement links with "Launch" actions that point to PDF files.
  - Draw borders around links.
  - Handle links with named destinations.
  - GoToR links specify a page number instead of a page reference.
Optimizations:

xpdf/pdftops.cc
//========================================================================
//
// pdftops.cc
//
// Copyright 1996 Derek B. Noonburg
//
//========================================================================

#include <aconf.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include "parseargs.h"
#include "GString.h"
#include "gmem.h"
#include "GlobalParams.h"
#include "Object.h"
#include "Stream.h"
#include "Array.h"
#include "Dict.h"
#include "XRef.h"
#include "Catalog.h"
#include "Page.h"
#include "PDFDoc.h"
#include "PSOutputDev.h"
#include "Error.h"
#include "config.h"

static int firstPage = 1;
static int lastPage = 0;
static GBool level1 = gFalse;
static GBool level1Sep = gFalse;
static GBool level2Sep = gFalse;
static GBool doEPS = gFalse;
static GBool doForm = gFalse;
#if OPI_SUPPORT
static GBool doOPI = gFalse;
#endif
static GBool noEmbedT1Fonts = gFalse;
static GBool noEmbedTTFonts = gFalse;
static char paperSize[15] = "";
static int paperWidth = 0;
static int paperHeight = 0;
static GBool duplex = gFalse;
static char ownerPassword[33] = "";
static char userPassword[33] = "";
static GBool quiet = gFalse;
static char cfgFileName[256] = "";
static GBool printVersion = gFalse;
static GBool printHelp = gFalse;

static ArgDesc argDesc[] = {
  {"-f",      argInt,      &firstPage,      0,
   "first page to print"},
  {"-l",      argInt,      &lastPage,       0,
   "last page to print"},
  {"-level1", argFlag,     &level1,         0,
   "generate Level 1 PostScript"},
  {"-level1sep", argFlag,  &level1Sep,      0,
   "generate Level 1 separable PostScript"},
  {"-level2sep", argFlag,  &level2Sep,      0,
   "generate Level 2 separable PostScript"},
  {"-eps",    argFlag,     &doEPS,          0,
   "generate Encapsulated PostScript (EPS)"},
  {"-form",   argFlag,     &doForm,         0,
   "generate a PostScript form"},
#if OPI_SUPPORT
  {"-opi",    argFlag,     &doOPI,          0,
   "generate OPI comments"},
#endif
  {"-noembt1", argFlag,     &noEmbedT1Fonts, 0,
   "don't embed Type 1 fonts"},
  {"-noembtt", argFlag,    &noEmbedTTFonts, 0,
   "don't embed TrueType fonts"},
  {"-paper",  argString,   paperSize,       sizeof(paperSize),
   "paper size (letter, legal, A4, A3)"},
  {"-paperw", argInt,      &paperWidth,     0,
   "paper width, in points"},
  {"-paperh", argInt,      &paperHeight,    0,
   "paper height, in points"},
  {"-duplex", argFlag,     &duplex,         0,
   "enable duplex printing"},
  {"-opw",    argString,   ownerPassword,   sizeof(ownerPassword),
   "owner password (for encrypted files)"},
  {"-upw",    argString,   userPassword,    sizeof(userPassword),
   "user password (for encrypted files)"},
  {"-q",      argFlag,     &quiet,          0,
   "don't print any messages or errors"},
  {"-cfg",        argString,      cfgFileName,    sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-v",      argFlag,     &printVersion,   0,
   "print copyright and version info"},
  {"-h",      argFlag,     &printHelp,      0,
   "print usage information"},
  {"-help",   argFlag,     &printHelp,      0,
   "print usage information"},
  {"--help",  argFlag,     &printHelp,      0,
   "print usage information"},
  {"-?",      argFlag,     &printHelp,      0,
   "print usage information"},
  {NULL}
};

int main(int argc, char *argv[]) {
  PDFDoc *doc;
  GString *fileName;
  GString *psFileName;
  PSLevel level;
  PSOutMode mode;
  GString *ownerPW, *userPW;
  PSOutputDev *psOut;
  GBool ok;
  char *p;

  // parse args
  ok = parseArgs(argDesc, &argc, argv);
  if (!ok || argc < 2 || argc > 3 || printVersion || printHelp) {
    fprintf(stderr, "pdftops version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("pdftops", "<PDF-file> [<PS-file>]", argDesc);
    }
    exit(1);
  }
  if ((level1 && level1Sep) ||
      (level1Sep && level2Sep) ||
      (level1 && level2Sep)) {
    fprintf(stderr, "Error: use only one of -level1, -level1sep, and -level2sep.\n");
    exit(1);
  }
  if (doEPS && doForm) {
    fprintf(stderr, "Error: use only one of -eps and -form\n");
    exit(1);
  }
  if (doForm && (level1 || level1Sep)) {
    fprintf(stderr, "Error: forms are only available with Level 2 output.\n");
    exit(1);
  }
  fileName = new GString(argv[1]);
  level = level1 ? psLevel1
                 : level1Sep ? psLevel1Sep
                             : level2Sep ? psLevel2Sep
                                         : psLevel2;
  mode = doEPS ? psModeEPS
               : doForm ? psModeForm
                        : psModePS;

  // read config file
  globalParams = new GlobalParams(cfgFileName);
  if (paperSize[0]) {
    if (!globalParams->setPSPaperSize(paperSize)) {
      fprintf(stderr, "Invalid paper size\n");
      exit(1);
    }
  } else {
    if (paperWidth) {
      globalParams->setPSPaperWidth(paperWidth);
    }
    if (paperHeight) {
      globalParams->setPSPaperHeight(paperHeight);
    }
  }
  if (duplex) {
    globalParams->setPSDuplex(duplex);
  }
  if (level1 || level1Sep || level2Sep) {
    globalParams->setPSLevel(level);
  }
  if (noEmbedT1Fonts) {
    globalParams->setPSEmbedType1(!noEmbedT1Fonts);
  }
  if (noEmbedTTFonts) {
    globalParams->setPSEmbedTrueType(!noEmbedTTFonts);
  }
#if OPI_SUPPORT
  if (doOPI) {
    globalParams->setPSOPI(doOPI);
  }
#endif
  if (quiet) {
    globalParams->setErrQuiet(quiet);
  }

  // open PDF file
  if (ownerPassword[0]) {
    ownerPW = new GString(ownerPassword);
  } else {
    ownerPW = NULL;
  }
  if (userPassword[0]) {
    userPW = new GString(userPassword);
  } else {
    userPW = NULL;
  }
  doc = new PDFDoc(fileName, ownerPW, userPW);
  if (userPW) {
    delete userPW;
  }
  if (ownerPW) {
    delete ownerPW;
  }
  if (!doc->isOk()) {
    goto err1;
  }

  // check for print permission
  if (!doc->okToPrint()) {
    error(-1, "Printing this document is not allowed.");
    goto err1;
  }

  // construct PostScript file name
  if (argc == 3) {
    psFileName = new GString(argv[2]);
  } else {
    p = fileName->getCString() + fileName->getLength() - 4;
    if (!strcmp(p, ".pdf") || !strcmp(p, ".PDF")) {
      psFileName = new GString(fileName->getCString(),
			       fileName->getLength() - 4);
    } else {
      psFileName = fileName->copy();
    }
    psFileName->append(doEPS ? ".eps" : ".ps");
  }

  // get page range
  if (firstPage < 1) {
    firstPage = 1;
  }
  if (lastPage < 1 || lastPage > doc->getNumPages()) {
    lastPage = doc->getNumPages();
  }

  // check for multi-page EPS or form
  if ((doEPS || doForm) && firstPage != lastPage) {
    error(-1, "EPS and form files can only contain one page.");
    goto err2;
  }

  // write PostScript file
  psOut = new PSOutputDev(psFileName->getCString(), doc->getXRef(),
			  doc->getCatalog(), firstPage, lastPage, mode);
  if (psOut->isOk()) {
    doc->displayPages(psOut, firstPage, lastPage, 72, 0, gFalse);
  }
  delete psOut;

  // clean up
 err2:
  delete psFileName;
 err1:
  delete doc;
  delete globalParams;

  // check for memory leaks
  Object::memCheck(stderr);
  gMemReport(stderr);

  return 0;
}
